fix: depreciate allowClientClassCreation defaulting to true

dblythy opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
<!-- Add a brief description of the issue this PR solves. -->

`allowClientClassCreation` should be changed in a future version to default as `false`

Related issue: #7156

### Approach
<!-- Add a description of the approach in this PR. -->

Uses deprecator 

### TODOs before merging
<!--
    Add TODOs that need to be completed before merging this PR.
    Delete suggested TODOs that do not apply to this PR.
-->

- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

mtrezza committed:
ci: fix node engine check (#7891)

dblythy committed:
fix: return correct response when revert is used in beforeSave (#7839)

semantic-release-bot committed:
chore(release): 5.2.1-alpha.1 [skip ci]

## [5.2.1-alpha.1](https://github.com/parse-community/parse-server/compare/5.2.0...5.2.1-alpha.1) (2022-03-26)

### Bug Fixes

* return correct response when revert is used in beforeSave ([#7839](https://github.com/parse-community/parse-server/issues/7839)) ([f63fb2b](https://github.com/parse-community/parse-server/commit/f63fb2b338c908f0e7a648d338c26b9daa50c8f2))

dblythy committed:
perf: reduce database operations when using the constant parameter in Cloud Function validation (#7892)

semantic-release-bot committed:
chore(release): 5.2.1-alpha.2 [skip ci]

## [5.2.1-alpha.2](https://github.com/parse-community/parse-server/compare/5.2.1-alpha.1...5.2.1-alpha.2) (2022-03-26)

### Performance Improvements

* reduce database operations when using the constant parameter in Cloud Function validation ([#7892](https://github.com/parse-community/parse-server/issues/7892)) ([48bd512](https://github.com/parse-community/parse-server/commit/48bd512eeb47666967dff8c5e723ddc5b7801daa))

github-actions[bot] committed:
feat: add MongoDB 5.1 compatibility (#7682)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.1 [skip ci]

# [5.3.0-alpha.1](https://github.com/parse-community/parse-server/compare/5.2.1-alpha.2...5.3.0-alpha.1) (2022-03-27)

### Features

* add MongoDB 5.1 compatibility ([#7682](https://github.com/parse-community/parse-server/issues/7682)) ([90155cf](https://github.com/parse-community/parse-server/commit/90155cf1680e5e0499b0000e071c6cb0ce3aef96))

mtrezza committed:
fix: security upgrade parse push adapter from 4.1.0 to 4.1.2 (#7893)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.2 [skip ci]

# [5.3.0-alpha.2](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.1...5.3.0-alpha.2) (2022-03-27)

### Bug Fixes

* security upgrade parse push adapter from 4.1.0 to 4.1.2 ([#7893](https://github.com/parse-community/parse-server/issues/7893)) ([ef56e98](https://github.com/parse-community/parse-server/commit/ef56e98ef65041b4d3b7b82cce3473269c27f6fd))

mtrezza committed:
feat: add MongoDB 5.2 support (#7894)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.3 [skip ci]

# [5.3.0-alpha.3](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.2...5.3.0-alpha.3) (2022-03-27)

### Features

* add MongoDB 5.2 support ([#7894](https://github.com/parse-community/parse-server/issues/7894)) ([6b4b358](https://github.com/parse-community/parse-server/commit/6b4b358f0842ae920e45652f5e8b2afebc6caf3a))

snyk-bot committed:
refactor: upgrade mongodb from 4.3.1 to 4.4.1 (#7906)

github-actions[bot] committed:
ci: bump environment (#7907)

Moumouls committed:
refactor: add missing schema definitions (#7917)

dblythy committed:
fix: custom database options are not passed to MongoDB GridFS (#7911)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.4 [skip ci]

# [5.3.0-alpha.4](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.3...5.3.0-alpha.4) (2022-04-04)

### Bug Fixes

* custom database options are not passed to MongoDB GridFS ([#7911](https://github.com/parse-community/parse-server/issues/7911)) ([a72b384](https://github.com/parse-community/parse-server/commit/a72b384f76137a3d83ffb69f65cb25aff1bbab4f))

dblythy committed:
fix: depreciate allowClientClassCreation defaulting to true

dblythy committed:
Update DEPRECATIONS.md

snyk-bot committed:
refactor: upgrade @apollo/client from 3.5.8 to 3.5.9 (#7921)

dependabot[bot] committed:
fix: security upgrade moment from 2.29.1 to 2.29.2 (#7931)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.5 [skip ci]

# [5.3.0-alpha.5](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.4...5.3.0-alpha.5) (2022-04-09)

### Bug Fixes

* security upgrade moment from 2.29.1 to 2.29.2 ([#7931](https://github.com/parse-community/parse-server/issues/7931)) ([6b68593](https://github.com/parse-community/parse-server/commit/6b68593eaec17e8b183899d2b92699c9ede7625b))

mtrezza committed:
Merge branch 'alpha' into depreciate-allow-client-class

mtrezza reviewed (commented):
Should this PR also change the default?

marvinroger committed:
fix: peer dependency mismatch for GraphQL dependencies (#7934)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.6 [skip ci]

# [5.3.0-alpha.6](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.5...5.3.0-alpha.6) (2022-04-11)

### Bug Fixes

* peer dependency mismatch for GraphQL dependencies ([#7934](https://github.com/parse-community/parse-server/issues/7934)) ([b7a1d76](https://github.com/parse-community/parse-server/commit/b7a1d7617b4bcac677cecedfeb6ac4a27447083b))

snyk-bot committed:
refactor: upgrade @graphql-tools/links from 8.2.2 to 8.2.4 (#7932)

snyk-bot committed:
refactor: upgrade @graphql-tools/links from 8.2.4 to 8.2.6 (#7935)

dplewis committed:
refactor: upgrade @graphql-tools/links from 8.2.6 to 8.2.7 (#7941)

dplewis committed:
refactor: upgrade body-parser from 1.19.1 to 1.19.2 (#7944)

dblythy commented:
Isn’t the point of depreciating that we give users a warning that the default will change in a future version?

mtrezza commented:
You are absolutely right, this looks fine. Re-running tests until they all pass.

mtrezza committed:
Merge branch 'alpha' into depreciate-allow-client-class

dblythy commented:
Actually, I still think I have it wrong. Shouldn’t it be depreciated in the next version (≈5.2) and default changed in version 6?

mtrezza commented:
Well, our policy is to give developers min. 1 major version period (effectively min. 1 year) time to adjust to a breaking change. If we change the default in v6 then we give only 8 months time. So I believe you got the PR right as it is now.

dplewis committed:
refactor: upgrade @graphql-tools/links from 8.2.7 to 8.2.8 (#7945)

mtrezza committed:
Merge branch 'alpha' into depreciate-allow-client-class

snyk-bot committed:
fix: security upgrade @parse/fs-files-adapter from 1.2.1 to 1.2.2 (#7948)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.7 [skip ci]

# [5.3.0-alpha.7](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.6...5.3.0-alpha.7) (2022-04-25)

### Bug Fixes

* security upgrade @parse/fs-files-adapter from 1.2.1 to 1.2.2 ([#7948](https://github.com/parse-community/parse-server/issues/7948)) ([20fc4e2](https://github.com/parse-community/parse-server/commit/20fc4e23b53c91aac657f894bd70d049b7525c37))

dblythy commented:
It just seems a bit wrong that the _actual_ change won't happen until 2024. Can we notify in the next version that change will happen in May 2023?

mtrezza commented:
What seems wrong about it and why would the change be in May 2023?

dblythy commented:
Well, not that it's _wrong_, it's just less than ideal to have to wait 18 months for a security improvement to actually be implemented

mtrezza commented:
It's not a bug but a config default change, so everyone can change this already with the current version of Parse Server, so I don't see an urgency here. Waiting for 18 months is just the trade-off to avoid frequent breaking changes. In my opinion, the fact that we got these previously unpredictable breaking changes in Parse Server under control now by far outweighs the additional 8 months waiting time. What do you think about maybe adding this to the security best practice guide and if you write up a short tweet we can tweet about the guide from our official twitter account as a reminder?

dblythy commented:
Sounds good - I guess I just thought the depreciation could happen now and the actual default change to happen in a version released in 12 months, rather than waiting til v7 to add the depreciation

dblythy committed:
Merge branch 'alpha' into depreciate-allow-client-class

mtrezza commented:
Oh, sure, we can add the deprecation warning already now, I thought you wanted to do that anyway as it's already in this PR. The deprecation has to be min. 1 year, but if it's longer, even better as it gives developers more time to prepare. So IMO we can merge this PR anytime.

dblythy commented:
Ohhh, right. I thought that depreciations could only happened in major versions. I’ll update the table

mtrezza committed:
refactor: upgrade winston from 3.5.1 to 3.7.2 (#7954)

dblythy committed:
Update DEPRECATIONS.md

dblythy committed:
Merge branch 'alpha' into depreciate-allow-client-class

dblythy commented:
@mtrezza let me know how that looks now. As there is a new item in the Deprecator/Deprecations.js, the option `allowClientClassCreation` is depreciated in the next release, but the default change won't happen for 12 months, until May 2023

mtrezza commented on DEPRECATIONS.md:
```diff
@@ -8,7 +8,7 @@ The following is a list of deprecations, according to the [Deprecation Policy](h
 | DEPPS3 | Config option `enforcePrivateUsers` defaults to `true` | [#7319](https://github.com/parse-community/parse-server/pull/7319)   | 5.0.0 (2022)                    | 6.0.0 (2023)                    | deprecated            | -     |
 | DEPPS4 | Remove convenience method for http request `Parse.Cloud.httpRequest` | [#7589](https://github.com/parse-community/parse-server/pull/7589)   | 5.0.0 (2022)                    | 6.0.0 (2023)                    | deprecated            | -     |
-| DEPPS5 | Config option `allowClientClassCreation` defaults to `false` | [#7925](https://github.com/parse-community/parse-server/pull/7925)   | 6.0.0 (2023)                    | 7.0.0 (2024)                    | deprecated            | -     |
+| DEPPS5 | Config option `allowClientClassCreation` defaults to `false` | [#7925](https://github.com/parse-community/parse-server/pull/7925)   | 5.3.0 (2023)                    | >6.0.0 (May 2024)                    | deprecated            | -     |
```
That can't be correct. We can begin deprecation anytime, as from `5.3.0 (2023)`, but the breaking change can only happen on a major version upgrade, otherwise we violate semver. So instead of `>6.0.0 (May 2024) ` it should be `7.0.0 (2024)`. In addition, the breaking change must happen at least 12 months from the beginning of deprecation, otherwise we violate our deprecation policy.

Maybe the terminology is causing confusion. "to deprecate" means that something should not be used anymore because it will change in a future version but it's still permissible to be used. at the end of the "deprecation period" the breaking change will happen, which can only happen on a major version increment as per semver. 

The concepts of _deprecation_ and _semver_ are distinct from each other.

dblythy replied:
Ok thank you for the explanation

dblythy committed:
Update DEPRECATIONS.md

mtrezza commented:
I think this is ready for merge now, what do you think? I'm just re-running the tests until they pass as a formality.

mtrezza committed:
Merge branch 'alpha' into depreciate-allow-client-class

mtrezza commented:
Resolved conflicts; if CI passes, this can be merged

mtrezza commented:
Some CI jobs are failing:

```
1) OAuth POST request for a resource that requires OAuth should fail with invalid credentials
- Expected '34' to equal '32'.
```

It fails only for some tests, re-running the tests makes them fail again, perhaps a Twitter API limiting issue; looks unrelated to this PR.

parseplatformorg commented:
🎉 This change has been released in version [5.3.0-alpha.9](https://github.com/parse-community/parse-server/releases/tag/5.3.0-alpha.9)

parseplatformorg commented:
🎉 This change has been released in version [5.3.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/5.3.0-beta.1)

parseplatformorg commented:
🎉 This change has been released in version [5.3.0](https://github.com/parse-community/parse-server/releases/tag/5.3.0)
